### PR TITLE
test: migrate all packages from Flutter driver to integration tests

### LINKED ISSUE #8909: Migrate e2e tests from Flutter Driver to `integration_test` package
## Description
The `integration_test` is the new way to write e2e tests. Therefore, we should migrate all packages from Flutter Driver to `integration_test`.

We will split the migration into multiple pull requests. All pull requests will be merged into the [`migrate-to-integration-tests`](https://github.com/firebase/flutterfire/tree/migrate-to-integration-tests) branch.

We should also check at the end if the CI is flaky by running like 100 times the same job.

### PR DESCRIPTION
**Pull request by nilsreichardt**
## Description

The `integration_test` package is the new way to write e2e tests. Therefore, this PR migrates the tests from driver tests to integration tests.

As side effect, we see that the Android CI e2e tests pipeline is way more stable.

https://user-images.githubusercontent.com/24459435/193446727-cc53f7a9-3725-4a29-a701-fe8a810bbe71.mov

### ToDos:

- [x] Migrate `cloud_firestore`
- [x] Migrate `cloud_functions`
- [x] Migrate `firebase_analytics`
- [x] Migrate `firebase_app_installations`
- [x] Migrate `firebase_auth`
- [x] Migrate `firebase_core`
- [x] Migrate `firebase_crashlytics`
- [x] Migrate `firebase_database`
- [x] Migrate `firebase_dynamic_links`
- [x] Migrate `firebase_in_app_messaging`
- [x] Migrate `firebase_ml_model_downloader`
- [x] Migrate `firebase_remote_config`
- [x] Migrate `firebase_messaging`
- [x] Migrate `firebase_storage`
- [x] Migrate `cloud_firestore_odm`
- [x] Migrate `firebase_performance`
- [x] Migrate `flutterfire_ui`
- [x] Delete driver tests (test_driver/ folder)
- [x] Delete unused `cloud_functions` Flutter driver tests (in `packages/cloud_functions/cloud_functions/example)`
- [x] Delete unused `firebase_analytics` Flutter driver tests (in `packages/firebase_analytics/firebase_analytics/example)`
- [x] Delete unused `firebase_auth` Flutter driver tests (in `packages/firebase_auth/firebase_auth/example)`
- [x] Delete unused `firebase_core` Flutter driver tests (in `packages/firebase_core/firebase_core/example)`
- [x] Remove unused dev dependencies
- [x] Remove `drive` as dependency
- [ ] Mark `drive` dependency on pub.dev as discontinued?
- [x] Check if CI is flaky (run CI 50 times)
- [x] Update documentation about running integration tests
- [x] Update Melos commands

## Related Issues

Closes #8909

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change.
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/firebase/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/cloud_functions/cloud_functions/pubspec.yaml
Comment (nilsreichardt) [inline packages/cloud_functions/cloud_functions/pubspec.yaml:28]: I removed the `test` dependencies in all packages (except `_flutterfire_internals`) because the packages already have `flutter_test` as a dependency. Or is there a reason why we have two test dependencies?

File: tests/android/app/build.gradle
Comment (nilsreichardt) [inline tests/android/app/build.gradle:54]: Required to build for Android. Otherwise, you are getting an error.

File: packages/flutterfire_ui/example/test_driver/integration_test.dart
Comment (nilsreichardt) [inline packages/flutterfire_ui/example/test_driver/integration_test.dart:3]: Are the tests of `flutterfire` used? They are not in the CI, right?

File: tests/integration_test/firebase_auth/firebase_auth_multi_factor_e2e_test.dart
Comment (nilsreichardt) [inline tests/integration_test/firebase_auth/firebase_auth_multi_factor_e2e_test.dart:237]: Typo fix

File: packages/firebase_auth/firebase_auth/example/test_driver/user_e2e.dart
Comment (nilsreichardt) [inline packages/firebase_auth/firebase_auth/example/test_driver/user_e2e.dart:1]: Removing the tests from the `firebase_auth` from `packages/firebase_auth/firebase_auth/example/` because the source of truth should be in `tests`, right?

File: tests/integration_test/cloud_functions/cloud_functions_e2e_test.dart
Comment (nilsreichardt) [inline tests/integration_test/cloud_functions/cloud_functions_e2e_test.dart:198]: Skipping this test because it's flaky. See: https://github.com/firebase/flutterfire/issues/9652

File: tests/integration_test/firebase_messaging/firebase_messaging_e2e_test.dart
Comment (nilsreichardt) [inline tests/integration_test/firebase_messaging/firebase_messaging_e2e_test.dart:204]: Android skipped because it times out with a likelihood of 98%. See: https://github.com/firebase/flutterfire/issues/9651

File: tests/integration_test/firebase_messaging/firebase_messaging_e2e_test.dart
Comment (nilsreichardt) [inline tests/integration_test/firebase_messaging/firebase_messaging_e2e_test.dart:219]: Android skipped because it times out with a likelihood of 98%. See: https://github.com/firebase/flutterfire/issues/9650

File: packages/flutterfire_ui/example/test_driver/integration_test.dart
Comment (russellwheatley) [reply packages/flutterfire_ui/example/test_driver/integration_test.dart:3]: We don't have integration tests for `flutterfire_ui` if that is your question 😄

File: packages/firebase_auth/firebase_auth/example/test_driver/user_e2e.dart
Comment (russellwheatley) [reply packages/firebase_auth/firebase_auth/example/test_driver/user_e2e.dart:1]: Makes sense to me. Double check with @Salakar as he set it up and might have a reason why e2e tests remain in example app.

**Review by russellwheatley — COMMENTED** [review]
LGTM provided a couple of questions I'm not sure on are answered by the rest of the core team. Good work, @nilsreichardt 🎉

File: packages/cloud_firestore_odm/cloud_firestore_odm/example/build.yaml
Comment (rrousselGit) [reply packages/cloud_firestore_odm/cloud_firestore_odm/example/build.yaml:19]: Without it, the code-generator may not run on the test_driver/integration folder

Comment (rrousselGit) [reply packages/cloud_firestore_odm/cloud_firestore_odm/example/build.yaml:19]: Why does this need to be removed to begin with?

**Comment by nilsreichardt** [comment]
> Would it be possible to break this PR down into a few smaller ones?
> 
> Reviewing changes on 123 files is not reasonably doable.

Half of the migration has already been reviewed in #9230 (see #8910, #8946, #8938, #8972, #8975, #8974, #8978, #8973, #8977). However, cherry-picking the commits in separate branches should be relatively easy because I committed in very small steps.

Is still preferred? Yes, I can do this 👍

File: packages/flutterfire_ui/example/test_driver/integration_test.dart
Comment (nilsreichardt) [reply packages/flutterfire_ui/example/test_driver/integration_test.dart:3]: But there are some tests in `example` folder of the `flutterfire_ui` 🤔 https://github.com/firebase/flutterfire/tree/master/packages/flutterfire_ui/example/test_driver

Comment (russellwheatley) [reply packages/flutterfire_ui/example/test_driver/integration_test.dart:3]: Huh, they aren't part of our CI e2e tests. I think you can leave them for now and we'll integrate them at some other point. Unless @lesnitsky thinks otherwise?

Comment (nilsreichardt) [reply packages/flutterfire_ui/example/test_driver/integration_test.dart:3]: Should we add them also? If yes, I can make a new issue and integrate them in a separate PR into our CI 👍

File: packages/cloud_firestore_odm/cloud_firestore_odm/example/build.yaml
Comment (rrousselGit) [reply packages/cloud_firestore_odm/cloud_firestore_odm/example/build.yaml:19]: Ah, good catch. Weird

That's fine then

File: packages/cloud_firestore_odm/cloud_firestore_odm/example/integration_test/cloud_firestore_odm_e2e_test.dart
Comment (bartekpacia) [inline packages/cloud_firestore_odm/cloud_firestore_odm/example/integration_test/cloud_firestore_odm_e2e_test.dart:16]: afaik this isn't required, because the flutter tool creates a temporary file where the call to `IntegrationTestWidgetsFlutterBinding.ensureInitialized()` is automatically inserted. See https://github.com/flutter/flutter/issues/92254 for more details :)

Comment (nilsreichardt) [reply packages/cloud_firestore_odm/cloud_firestore_odm/example/integration_test/cloud_firestore_odm_e2e_test.dart]: Thanks for the note, @bartekpacia. However, it seems that without `IntegrationTestWidgetsFlutterBinding.ensureInitialzed()` the tests doesn't on web. Maybe it's related with https://github.com/flutter/flutter/issues/66264

See:
* https://github.com/firebase/flutterfire/actions/runs/3213776146/jobs/5253698990
* https://github.com/firebase/flutterfire/actions/runs/3213776149/jobs/5253699165
* https://github.com/firebase/flutterfire/actions/runs/3213776150/jobs/5253699260

**Review by Lyokone — APPROVED** [review]

**Comment by Lyokone** [comment]
Reviewed carefully and everything seems fine. Some files are deleted because they were duplicates not being used anymore. LGTM

File: CONTRIBUTING.md
Comment (Lyokone) [inline CONTRIBUTING.md:123]: You have the `melos run firebase:emulator` than can be used

**Review by Salakar — APPROVED** [review]

**Pull request merged**
